go/mysql/mysql56_gtid_set.go

```diff
@@ -334,6 +334,83 @@ func (set Mysql56GTIDSet) AddGTID(gtid GTID) GTIDSet {
 	return newSet
 }
 
+// Union implements GTIDSet.Union().
+func (set Mysql56GTIDSet) Union(other GTIDSet) GTIDSet {
+	if set == nil || other == nil {
+		return set
+	}
+	mydbOther, ok := other.(Mysql56GTIDSet)
+	if !ok {
+		return set
+	}
+
+	// Make a copy and add the new GTID in the proper place.
+	// This function is not supposed to modify the original set.
+	newSet := make(Mysql56GTIDSet)
+
+	for OtherSID, otherIntervals := range mydbOther {
+		intervals, ok := set[OtherSID]
+
+		if !ok {
+			// No matching server id, so we must add it from other set.
+			newSet[OtherSID] = otherIntervals
+			continue
+		}
+
+		// Found server id match between sets, so now we need to add each interval.
+		s1 := intervals[:]
+		s2 := otherIntervals[:]
+		var nextInterval interval
+		var newIntervals []interval
+
+		// While our stacks have intervals to process, do work.
+		for len(s1) != 0 || len(s2) != 0 {
+			// Find which intervals list has earliest start.
+			if len(s2) == 0 || (len(s1) != 0 && s1[0].start <= s2[0].start) {
+				nextInterval = s1[0]
+				// Progress pointer since this stack has earliest.
+				s1 = s1[1:]
+			} else {
+				nextInterval = s2[0]
+				// Progress pointer since this stack has earliest.
+				s2 = s2[1:]
+			}
+
+			if len(newIntervals) == 0 {
+				newIntervals = append(newIntervals, nextInterval)
```

Comment:
Shouldn't we `continue` after this?

Reply:
Ahh, that's true yes. I wish there was an easy way to do it out of the while loop entirely, but couldn't think of a way without duplicating the popping logic into two places. Once I add the pop helper function how do you feel about doing an initial pop outside the while loop into the intervals list rather than having this in the while loop at all?

Reply:
I'm ok with either inside the loop or outside.

Reply:
Decided to just continue. Because the nature of the pop helper method returns a bool, rather than the popped item, it was cleaner to just continue here instead of pre-popping.